x-pack/plugins/enterprise_search/public/applications/app_search/components/crawler/components/add_domain/add_domain_logic.ts

```diff
@@ -0,0 +1,164 @@
+        setAddDomainFormInputValue: () => '/',
+        onValidateDomain: (_, { newEntryPointValue }) => newEntryPointValue,
+      },
+    ],
+    // TODO When 4-step validation is added this will become a selector as
+    // we'll use individual step results to determine whether this is true/false
+    hasValidationCompleted: [
+      false,
+      {
+        clearDomainFormInputValue: () => false,
+        setAddDomainFormInputValue: () => false,
+        onValidateDomain: () => true,
+      },
+    ],
+    errors: [
+      [],
+      {
+        clearDomainFormInputValue: () => [],
+        setAddDomainFormInputValue: () => [],
+        onValidateDomain: () => [],
+        submitNewDomain: () => [],
+        onSubmitNewDomainError: (_, { errors }) => errors,
+      },
+    ],
+  }),
+  selectors: ({ selectors }) => ({
+    // TODO include selectors.blockingFailures once 4-step validation is migrated
+    allowSubmit: [
+      () => [selectors.hasValidationCompleted], // should eventually also contain selectors.hasBlockingFailures when that is added
+      (hasValidationCompleted: boolean) => hasValidationCompleted, // && !hasBlockingFailures
+    ],
+  }),
+  listeners: ({ actions, values }) => ({
+    onSubmitNewDomainSuccess: ({ domain }) => {
+      flashSuccessToast(
+        i18n.translate(
+          'xpack.enterpriseSearch.appSearch.crawler.domainsTable.action.add.successMessage',
+          {
+            defaultMessage: "Successfully added domain '{domainUrl}'",
+            values: {
+              domainUrl: domain.url,
+            },
+          }
+        )
+      );
+      KibanaLogic.values.navigateToUrl(
+        generateEnginePath(ENGINE_CRAWLER_DOMAIN_PATH, { domainId: domain.id })
+      );
+    },
+    submitNewDomain: async () => {
+      const { http } = HttpLogic.values;
+      const { engineName } = EngineLogic.values;
+
+      const requestBody = JSON.stringify({
+        name: values.addDomainFormInputValue.trim(),
+        entry_points: [{ value: values.entryPointValue }],
+      });
+
+      try {
+        const response = await http.post(`/api/app_search/engines/${engineName}/crawler/domains`, {
+          query: {
+            respond_with: 'crawler_details',
+          },
+          body: requestBody,
+        });
+
+        const crawlerData = crawlerDataServerToClient(response as CrawlerDataFromServer);
+        CrawlerOverviewLogic.actions.onReceiveCrawlerData(crawlerData);
+        const newDomain = crawlerData.domains[crawlerData.domains.length - 1];
+        if (newDomain) {
+          actions.onSubmitNewDomainSuccess(newDomain);
+        }
+        // If there is not a new domain, that means the server responded with a 200 but
+        // didn't actually persist the new domain to our BE, and we take no action
+      } catch (e) {
+        // we surface errors inside the form instead of in flash messages
+        const errorMessages = Array.isArray(e?.body?.attributes?.errors)
```

Comment:
Is this code that reaches down into the response to parse out the errors something we usually don't have to deal with when we're using our flash message helpers? If so, it'd be awesome if that were decoupled from flash messages so we could take advantage of it here.

Reply:
Optional, btw. Not suggesting you need to do that in this PR.

Reply:
I did this in ea57821

Reply:
I like that!